sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala

```diff
@@ -230,11 +232,29 @@ private[spark] class HiveExternalCatalog(conf: SparkConf, hadoopConf: Configurat
     // specify location for managed table. And in [[CreateDataSourceTableAsSelectCommand]] we have
     // to create the table directory and write out data before we create this table, to avoid
     // exposing a partial written table.
-    val needDefaultTableLocation = tableDefinition.tableType == MANAGED &&
-      tableDefinition.storage.locationUri.isEmpty
-
-    val tableLocation = if (needDefaultTableLocation) {
-      Some(CatalogUtils.stringToURI(defaultTablePath(tableDefinition.identifier)))
+    //
+    // When using a remote metastore, and if a managed table is being created with its
+    // location explicitly set to the location where it would be created anyway, then do
+    // not set its location explicitly. This avoids an issue with Sentry in secure clusters.
+    // Otherwise, the above comment applies.
```

Comment:
We have comments saying
```
// We can't leave `locationUri` empty and count on Hive metastore to set a default table
// location, because Hive metastore uses hive.metastore.warehouse.dir to generate default
// table location for tables in default database, while we expect to use the location of
// default database.
```

In general I think it's OK to always set the location for managed tables, and I feel a little weird if some systems forbid it. Hive supports setting a location for managed tables, how shall we handle that?

Reply:
I'm not sure I really follow that comment. IIRC you can't change the location of the default database in Hive, so I'm not sure how you'd hit a situation where "default db uri != warehouse dir", at least when using a remote metastore.

Reply:
> In general I think it's OK to always set the location for managed tables

BTW I don't disagree with that and as I said I believe this is ultimately a bug in Sentry, not in Spark. The question is whether we want this workaround in Spark while Sentry is not fixed.

Reply:
> IIRC you can't change the location of the default database in Hive

good to know that, maybe we should apply the same rule to Spark.

Reply:
>  The question is whether we want this workaround in Spark while Sentry is not fixed.
We could open a bug for Sentry and since Spark uses hive which often comes with Sentry in enterprise envs then it would be good to point to that bug in the code , so when things get fixed we could change bug the code here.